src/main/scala/mesosphere/marathon/core/storage/store/impl/cache/LazyCachingPersistenceStore.scala

```diff
@@ -159,3 +162,190 @@ class LazyCachingPersistenceStore[K, Category, Serialized](
 
   override def toString: String = s"LazyCachingPersistenceStore($store)"
 }
+
+case class LazyVersionCachingPersistentStore[K, Category, Serialized](
+    store: PersistenceStore[K, Category, Serialized],
+    config: VersionCacheConfig = VersionCacheConfig.Default)(implicit
+  mat: Materializer,
+    ctx: ExecutionContext) extends PersistenceStore[K, Category, Serialized] with StrictLogging {
+
+  private val lockManager = LockManager.create()
+  private[store] val versionCache = TrieMap.empty[(Category, K), Seq[OffsetDateTime]]
+  private[store] val versionedValueCache = TrieMap.empty[(K, OffsetDateTime), Option[Any]]
+
+  def withVersionedValueCache[Id, V, T](
+    future: => Future[T])(implicit ir: IdResolver[Id, V, Category, K]): Future[T] =
+    lockManager.executeSequentially(s"versionedValueCache::${ir.category}")(future)
+
+  def withVersionCache[Id, V, T](
+    id: Id)(future: (Category, K) => Future[T])(implicit ir: IdResolver[Id, V, Category, K]): Future[T] = {
+    val category = ir.category
+    val storageId = ir.toStorageId(id, None)
+    lockManager.executeSequentially((category, storageId).toString())(future(category, storageId))
+  }
+
+  protected[cache] def maybePurgeCachedVersions(
```

Comment:
we could. @jasongilanfarr suggest a random-purge approach so that's what I implemented. happy to implement LRU if we think it's going to work out better.